pkg/segment/query/processor/searcher.go

```diff
@@ -149,6 +158,113 @@ func (s *Searcher) Fetch() (*iqr.IQR, error) {
 	}
 }
 
+func (s *Searcher) fetchColumnSortedRRCs() (*iqr.IQR, error) {
+	qsrs, err := query.GetSortedQSRs(s.queryInfo, s.startTime, s.querySummary)
+	if err != nil {
+		log.Errorf("qid=%v, searcher.getBlocks: failed to get sorted QSRs: %v", s.qid, err)
+		return nil, err
+	}
+
+	// TODO: read PQS if enabled.
+
+	result := iqr.NewIQR(s.queryInfo.GetQid())
+	sorter := &sortProcessor{
+		options: s.sortExpr,
+	}
+
+	for _, qsr := range qsrs {
+		nextIQR, err := s.fetchSortedRRCsForQSR(qsr)
+		if err != nil {
+			log.Errorf("qid=%v, searcher.fetchColumnSortedRRCs: failed to fetch sorted RRCs: %v", s.qid, err)
+			return nil, err
+		}
+
+		result, err = iqr.MergeAndDiscardAfter(result, nextIQR, sorter.less, s.sortExpr.Limit)
+		if err != nil {
+			log.Errorf("qid=%v, searcher.fetchColumnSortedRRCs: failed to merge IQRs: %v", s.qid, err)
+			return nil, err
+		}
+	}
+
+	// TODO: don't always return EOF.
+	return result, io.EOF
+}
+
+func (s *Searcher) fetchSortedRRCsForQSR(qsr *query.QuerySegmentRequest) (*iqr.IQR, error) {
+	// TODO: handle subsequent fetches to this QSR.
+
+	const recordLimit = 1000 // TODO: find a better way to limit.
+	cname := s.sortExpr.SortEles[0].Field
+	lines, err := sortindex.ReadSortIndex(qsr.GetSegKey(), cname, recordLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(lines) == 0 {
+		// TODO: raw search this segment if it has the cname
+		panic("got no results in sortindex")
```

Comment:
I guess this panic will get replaced with something else in another PR very soon ?

Reply:
Oops, I'll remove that panic. Yes, this will get handled properly in a few days